packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js

```diff
@@ -41,88 +169,114 @@ export function buildRootObject(vatPowers) {
+  function finishMeter({ state, self }) {
+    meterByID.init(
+      state.meterID,
+      // harden({ meter: self, updater: state.updater }), // XXX RESTORE
+      harden({ meter: self }), // XXX TEMP
+    );
+    meterIDByMeter.set(self, state.meterID);
+    meterNotifiersTemp.set(self, makeNotifierKit()); // XXX TEMP
+  }
 
-    const [doneP, doneRR] = producePRR();
-    running.set(vatID, doneRR);
-    doneP.catch(() => {}); // shut up false whine about unhandled rejection
+  function insistServiceAvailable() {
+    assert(!quiescentP, 'service not available, try again later');
+  }
+
+  async function upgradeVat(vatID, readyP, bundleID, options = {}) {
+    assert(vatAdminDev, `vatAdmin device not configured`);
+    const { vatParameters, upgradeMessage = 'vat upgraded', ...rest } = options;
+    const leftovers = Object.keys(rest);
+    if (leftovers.length) {
+      const bad = leftovers.join(',');
+      assert.fail(X`upgrade() received unknown options: ${bad}`);
+    }
+    assert.typeof(upgradeMessage, 'string', 'upgradeMessage not a string');
+    if (readyP) {
+      // eslint-disable-next-line @jessie.js/no-nested-await
+      await readyP.catch(() => true);
+    }
+    if (vatID === vatAdminVatID) {
+      baggage.init('doingSelfUpgrade', true);
+    }
+    const upgradeID = D(vatAdminDev).upgradeVat(
+      vatID,
+      bundleID,
+      vatParameters,
+      upgradeMessage,
+    );
+    const [upgradeCompleteP, upgradeRR] = producePRR();
+    pendingUpgrades.set(upgradeID, upgradeRR);
+    return upgradeCompleteP;
+  }
+
+  function assertRunningVat(vatID, operation, exitingOK) {
+    assert(
+      runningVats.has(vatID) && (exitingOK || !exitingVats.has(vatID)),
+      `attempt to perform ${operation}() on non-running vat ${vatID}`,
+    );
+  }
 
-    const adminNode = Far('adminNode', {
-      terminateWithFailure(reason) {
-        D(vatAdminNode).terminateWithFailure(vatID, reason);
+  const adminNodeKindHandle = provide(baggage, 'adminNode', () =>
+    makeKindHandle('adminNode'),
+  );
+  const makeAdminNode = defineDurableKind(
+    adminNodeKindHandle,
+    vatID => ({ vatID }),
+    {
+      terminateWithFailure({ state }, reason) {
+        insistServiceAvailable();
+        assertRunningVat(state.vatID, 'terminateWithFailure');
+        exitingVats.add(state.vatID);
+        D(vatAdminDev).terminateWithFailure(state.vatID, reason);
       },
-      done() {
+      done({ state }) {
+        // In some world it might make sense to allow this operation on
+        // terminated vats (since they *do* have a 'done' status after all,
+        // i.e., non-running) but we expect the promise that's returned here to
+        // reject if the vat fails, and if the vat has already exited then its
+        // exit status has been lost.  However, by rejecting the 'done'
+        // operation itself we can signal the non-running condition without
+        // promoting the illusion that we know whether it exited cleanly or not.
```

Comment:
Oh, you know, maybe we should store the exit status in the admin facet `state`, so it's available for folks calling `done` much later. In the normal case, callers will drop the admin facet once they know the vat is dead (or unneeded), which would allow the exit status data to be dropped too.

This is how Linux does it: the process table entry must stick around (in the "D" dead or "Z" zombie state) until the parent process reaps the child (and learns the exit code). The kernel retains responsibility for remembering the exit code until the parent asks about it.

Let's see, that would need a mapping from `vatID` to a separate per-adminNode facet (so `makeAdminNode` becomes a defineDurableKindMulti) that's used to save the exit status. Then `done` can reliably report vat-terminate-happy by checking `state.exited` instead of `assertRunningVat`. The vatID-to-reportExitStatusNode mapping would be annoying, though, it'd be a durable Map that needs to be populated during `newVatCallback` and deleted during `vatTerminated`. Maybe that wouldn't be too bad.

Reply:
> maybe we should store the exit status in the admin facet `state`

That's an attractive notion, but there's no graceful place to stand to do that consistently -- the exit status is signaled by resolving or rejecting the `done` promise, but that vatAdmin vat can't hold onto that across upgrades.  Upon upgrade the `done` promise is severed and a new one created, so the actual act of resolution/rejection is handled external to admin node of the vat that whose `done` promise it is.  In principle we could make the admin node object multifaceted, with one fact that gets exported to the client and one that is retained internally (by storing it in the `runningVats` table, say), and then have a method on the internal facet to inject a done state into it.  But (1) this is a lot of gymnastics to code for a relatively minor value add, and (2) it still doesn't free any external party who is really interested in watching for termination from the obligation to look for the severed `done` promise and re-request a new one when that happens, in which case they already have everything they need to know what's going on.

Reply:
Hm, I guess those interested parties would react to the first `done()` promise getting upgrade-disconnected by going back to the admin facet and requesting a new `done()` promise, and if the vat had terminated-happy, that second promise could be fulfilled as expected (thanks to the stashed `state.exited`).

The original `done` promise is lost, of course, but a new one can still be created and fulfilled correctly.